File: packages/quran_sessions/lib/src/domain/policies/session_join_window_policy.dart
import 'platform_scheduling_policy.dart';
import 'staging_qa_join_window_bypass.dart';

/// When RTC join credentials may be prefetched before the user taps Join.
class SessionJoinWindowPolicy {
  const SessionJoinWindowPolicy({
    this.prefetchLeadTime = PlatformSchedulingPolicy.joinWindowLeadTime,
    this.distribution = stagingQaJoinWindowBypassDefaultDistribution,
  });

  final Duration prefetchLeadTime;
  final String distribution;

  /// True when [now] is inside the join window (Q-VC-03: 15m before until endsAt).
  ///
  /// When [qaBypassUserId] is a staging QA account, window timing is skipped.
  bool isWithinJoinWindow({
    required DateTime startsAt,
    required DateTime endsAt,
    required DateTime now,
    String? qaBypassUserId,
  }) {
    if (isQaJoinWindowBypassEligible(
      userId: qaBypassUserId,
      distribution: distribution,
    )) {
      return true;
    }

    final windowStart = startsAt.subtract(prefetchLeadTime);
    return !now.isBefore(windowStart) && !now.isAfter(endsAt);
  }
}

File: functions/src/quranSessions/sessionJoinWindowPolicy.ts
import { JOIN_WINDOW_LEAD_MS } from "./platformSchedulingPolicy";
import { isQaJoinWindowBypassEligible } from "./stagingQaJoinWindowBypass";

/** Q-VC-03: join allowed from 15m before startsAt until endsAt. */
export function isWithinJoinWindow(params: {
  startsAt: Date;
  endsAt: Date;
  now: Date;
  leadTimeMs?: number;
}): boolean {
  const lead = params.leadTimeMs ?? JOIN_WINDOW_LEAD_MS;
  const windowStart = params.startsAt.getTime() - lead;
  const nowMs = params.now.getTime();
  return nowMs >= windowStart && nowMs <= params.endsAt.getTime();
}

/** Join window check with staging-only QA uid bypass (window timing only). */
export function isWithinJoinWindowOrQaBypass(params: {
  startsAt: Date;
  endsAt: Date;
  now: Date;
  leadTimeMs?: number;
  uid?: string;
}): boolean {
  if (isQaJoinWindowBypassEligible(params.uid)) {
    return true;
  }
  return isWithinJoinWindow({
    startsAt: params.startsAt,
    endsAt: params.endsAt,
    now: params.now,
    leadTimeMs: params.leadTimeMs,
  });
}
